src/Endpoint/ReadWrite.php

```diff
@@ -0,0 +1,185 @@
+
+namespace Nexcess\Sdk\Endpoint;
+
+use Nexcess\Sdk\ {
+  Endpoint\Read,
+  Endpoint\ReadWritable,
+  Exception\ApiException,
+  Exception\SdkException,
+  Model\Modelable as Model,
+  Response
+};
+
+/**
+ * Represents a writable API endpoint.
+ */
+abstract class ReadWrite extends Read implements ReadWritable {
+
+  /** @var callable Queued callback for wait(). */
+  protected $_wait_until;
+
+  /**
+   * {@inheritDoc}
+   */
+  public function create(array $data) : Model {
+    $model = static::MODEL;
+
+    return (new $model())->sync(
+      $this->_client
+        ->request('POST', static::ENDPOINT, ['json' => $data])
+        ->toArray()
+    );
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function delete($model_or_id) : ReadWritable {
+    $model = is_int($model_or_id) ?
+      $this->getModel($model_or_id) :
+      $model_or_id;
+    $this->_checkModelType($model);
+
+    $id = $model->offsetGet('id');
+    if (! is_int($id)) {
+      throw new ApiException(
+        ApiException::MISSING_ID,
+        ['model' => static::class]
+      );
+    }
+
+    $this->_client->request('DELETE', static::ENDPOINT . "/{$id}");
+    $this->_wait($this->_waitUntilDelete($model));
+
+    return $this;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function update(Model $model, array $data = []) : ReadWritable {
+    $this->_checkModelType($model);
+
+    $id = $model->offsetGet('id');
+    if (! $id) {
+      throw new ApiException(
+        ApiException::MISSING_ID,
+        ['model' => static::MODEL]
+      );
+    }
+
+    foreach ($data as $key => $value) {
+      $model->offsetSet($key, $value);
+    }
+
+    $update = empty($this->_stored[$id]) ?
+      $model->toArray() :
+      array_udiff_assoc(
+        $model->toArray(true),
+        $this->_stored[$id],
+        function ($value, $stored) { return ($value === $stored) ? 0 : 1; }
+      );
+
+    if (! empty($update)) {
+      return $this->_sync(
+        $this->_client
+          ->request('PATCH', static::ENDPOINT . "/{$id}/edit", $update)
+          ->toArray()
+      );
+    }
+
+    return $this;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function wait(
+    callable $until = null,
+    array $opts = []
+  ) : ReadWritable {
+    $config = $this->_config;
+
+    $until = $until ??
+      $this->_wait_until ??
+      function () { return true; };
+    $this->_wait_until = null;
+
+    $tick = $config->get('wait.tick_function');
+
+    $wait = $opts[self::OPT_WAIT_INTERVAL] ??
+      $config->get('wait.interval') ??
+      self::DEFAULT_WAIT_INTERVAL;
+
+    $timeout = $opts[self::OPT_WAIT_TIMEOUT] ??
+      $config->get('wait.timeout') ??
+      self::DEFAULT_WAIT_TIMEOUT;
+    $deadline = time() + $timeout;
+
+    try {
+      while ($until($this) !== true) {
+        if (time() > $deadline) {
+          throw new SdkException(
+            SdkException::WAIT_TIMEOUT_EXCEEDED,
+            ['timeout' => $timeout]
+          );
+        }
+
+        if ($tick) {
```

Comment:
you need to check if this is callable

Reply:
the Config class validated this already.

Reply:
https://github.com/nexcess/nexcess-php-sdk/pull/2/files#diff-74f8d3050ec63bf16014a5d32885e42bR63